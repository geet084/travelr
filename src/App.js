import React, { Component } from 'react';
import './Main.scss';

class App extends Component {
  render() {
    return (
      <div className="App">
        <header className="App-header">
          TRAVELR
        </header>
      </div>
    );
  }
}

export default App;