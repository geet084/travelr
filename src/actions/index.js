
export const isLoading = (bool) => ({
  type: 'IS_LOADING',
  isLoading: bool
})

export const hasErrored = (message) => ({
  type: 'HAS_ERRORED',
  message
})

export const fetchApodSuccess = (content) => ({
  type: 'FETCH_APOD_SUCCESS',
  content
})

export const fetchPlanetsSuccess = (planets) => ({
  type: 'FETCH_PLANETS_SUCCESS',
  planets
})

export const fetchImagesSuccess = (images) => ({
  type: 'FETCH_IMAGES_SUCCESS',
  images
})

export const setArrivalTime = (arrivalTime) => ({
  type: 'SET_ARRIVAL_TIME',
  arrivalTime
})

export const setBodies = (bodies) => ({
  type: 'SET_BODIES',
  bodies
})

export const setUserInfo = ({userDate, elapsedDays}) => ({
  type: 'SET_USER_INFO',
  userDate,
  elapsedDays,
})