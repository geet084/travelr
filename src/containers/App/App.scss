.App {
  text-align: center;
  display: flex;

}

nav {
  display: flex;
  width: 200px; 
  background-color: lightgrey;
}

.apod-img {
  position: absolute;
  z-index: -1;
}