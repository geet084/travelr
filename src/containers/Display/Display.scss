.display {
  color: white;
  margin-left: 200px;
  padding: 50px;
  font-size: 2rem;
}

.imgs {
  position: fixed;
  top: 0;
  left: 200px;
  margin: auto;
  height: 100%;
  z-index: -1;
}