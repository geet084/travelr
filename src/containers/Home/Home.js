import React, { Component } from 'react';
import DateForm from '../../components/DateForm/DateForm';
import PropTypes from 'prop-types';
import moment from 'moment';
import { connect } from 'react-redux';
import UserDate from '../../components/UserDate/UserDate';
import Distances from '../../components/Distances/Distances';
import Counter from '../../components/Counter/Counter';

export class Home extends Component {
  constructor() {
    super();
    this.state = {
      today: '',
      userDate: '',
      elapsedTime: 0,
      timeInSeconds: 0,
      earthSpin: 0,
      earthOrbit: 0,
      solarSystemOrbit: 0,
      galaxyMovement: 0,
      totalMovement: 0,
    }
  }

  componentDidMount = () => {
    this.calculateDateAndTime();
  }

  calculateDateAndTime = () => {
    const { arrivalTime } = this.props;
    const now = moment();

    const elapsedTimeSinceArrival = now.diff(arrivalTime);
    const timeInSeconds = elapsedTimeSinceArrival / 1000

    this.setState({
      timeInSeconds,
      today: now,
      earthSpin: (timeInSeconds * .28),
      earthOrbit: (timeInSeconds * 18.5),
      solarSystemOrbit: (timeInSeconds * 124.45),
      galaxyMovement: (timeInSeconds * 70),
      totalMovement: (timeInSeconds * 213.23),
    });
  }

  updateUserDate = (inputDate) => {
    // TODO: set time to a local noon or midnight?
    const { arrivalTime } = this.props;
    const userDate = moment(inputDate, 'MM-DD-YYYY');
    const elapsedTime = moment(arrivalTime).diff(userDate);
    
    this.setState({
      userDate,
      elapsedTime
    });
  }

  render() {
    const { url } = this.props;
    let { userDate, today, timeInSeconds, earthSpin, earthOrbit, solarSystemOrbit, galaxyMovement, totalMovement, } = this.state;

    const hasDate = today !== ''

    const days = hasDate ? today.diff(userDate, 'days') : 0
    const hours = hasDate ? today.diff(userDate, 'hours') - (days * 24) : 0
    const todaysDate = hasDate ? today.format('llll') : '';

    const total = ((days + (hours / 24)) * 18424000);
    const num = <Counter start={(total + totalMovement)} end={(total + totalMovement + 213230000)} decimal={1} />;
    
    return (
      <div className='home'>
        <img className='apod-img' src={url} alt="apod" />
        <header>
          <div className="prompt">
            <h3>How far do you think you've traveled today?</h3>
            <h3>How far have you traveled this week?</h3>
            <h3>Today is: {todaysDate}</h3>
            <h3>
              Enter a date to find out:
              <DateForm key={userDate} updateUserDate={this.updateUserDate} />
            </h3>
            {userDate !== '' && <UserDate userDate={userDate.format('llll')} days={days} hours={hours} num={num} />}
          </div>
        </header>
        {
          this.state.userDate !== '' &&
          <Distances
            timeInSeconds={timeInSeconds}
            earthSpin={earthSpin}
            earthOrbit={earthOrbit}
            solarSystemOrbit={solarSystemOrbit}
            galaxyMovement={galaxyMovement}
            totalMovement={totalMovement}
          />
        }
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  arrivalTime: state.arrivalTime,
})

Home.propTypes = {
  arrivalTime: PropTypes.object,
  url: PropTypes.string,
}

Home.defaultProps = {
  arrivalTime: {},
  url: '',
}

export default connect(mapStateToProps)(Home);