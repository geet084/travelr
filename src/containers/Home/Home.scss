.home {
  align-items: center;
  display: flex;
  justify-content: center;
  flex-direction: column;
  color: white;
  margin-left: 200px;
  margin: 0 auto;
}

.apod-img {
  height: 100%;
  position: fixed;
  left: 50%;
  top: 0;
  transform: translateX(-50%);
  z-index: -1;
}

header {
  display: flex;
  justify-content: space-between;
  padding: 10px 0 0 0;
}

.prompt {
  text-align: left;
  h3 {
    background-color: rgba(0, 0, 0, 0.7);
    box-shadow: 0px 0px 5px 5px rgba(0, 0, 0, 0.7);
  }
}