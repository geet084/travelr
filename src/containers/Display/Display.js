import React, { Component } from 'react';
import { connect } from 'react-redux';
import { handleImages } from '../../thunks/handleImages';
import { getImageSuccess } from '../../actions/';
import PropTypes from 'prop-types';

export class Display extends Component {

  componentDidMount = () => {
    let { info, handleImages } = this.props;
    const url = `https://images-api.nasa.gov/asset/${info.images[0]}`
    
    if (info.images !== []) handleImages(url, getImageSuccess)
  }

  render() {
    const { info } = this.props;
    const { currentImage } = this.props.images;
    
    return (
      <div className='display'>
        {info && <p>{'Name  - ' + info.object_name}</p>}
        {info && <p>{'Avg. Temp  - ' + info.average_temp}</p>}
        {info && <p>{'Distance from the sun  - '}</p>}
        {info && <p>{info.distance_from_sun}</p>}
        {info && <p>{'Length of day  - ' + info.length_of_day}</p>}
        {info && <p>{'Orbital period  - ' + info.orbital_period}</p>}

        {currentImage.href !== '' && <img className='imgs' src={currentImage.href} alt="" />}
      </div>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  handleImages: (url, actionToDispatch) => dispatch(handleImages(url, actionToDispatch))
})

export const mapStateToProps = (state) => ({
  images: state.images,
})

Display.propTypes = {
  handleImages: PropTypes.func,
  images: PropTypes.object,
  info: PropTypes.object,
}

Display.defaultProps = {
  images: {},
  info: {},
}

export default connect(mapStateToProps, mapDispatchToProps)(Display);