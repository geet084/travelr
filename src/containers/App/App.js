import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter, Route, Switch } from 'react-router-dom';
import { handleObjects } from '../../thunks/handleObjects';
import { handleImages } from '../../thunks/handleImages';
import { setArrivalTime, handleApodImage, handleObjectImages } from '../../actions'
import '../../Main.scss';
import Display from '../Display/Display';
import NavBar from '../../components/NavBar/NavBar'
import Home from '../../components/Home/Home'
import NotFound from '../../components/NotFound/NotFound'
import backupUrl from '../../images/back-img.jpg';
import PropTypes from 'prop-types';

export class App extends Component {

  componentDidMount = () => {
    const corsPrefix = 'https://cors-anywhere.herokuapp.com/'
    const apiKey = process.env.REACT_APP_NASA_APIKEY;
    const nasaURL = `${corsPrefix}https://api.nasa.gov/planetary/apod?api_key=${apiKey}`
    const serverURL = process.env.REACT_APP_SERVER_URL;

    this.props.handleImages(nasaURL, handleApodImage)
    this.props.handleObjects(serverURL + '/objects')
    this.props.handleImages(serverURL + '/images', handleObjectImages)
    this.props.setArrivalTime(Date.now())
  }

  render() {
    const { arrivalTime, objects, userInfo } = this.props;
    const { media_type, url } = this.props.images.apod;
    let currentUrl = url
    if (media_type === 'video' || url === undefined) currentUrl = backupUrl

    const planets = objects.filter(obj => obj.planet).map(obj => obj.object_name)
    const moons = objects.filter(obj => obj.moon).map(obj => obj.object_name)
    const bodies = objects.filter(obj => !obj.planet && !obj.moon).map(obj => obj.object_name)

    return (
      <div className="App">
        <h1 className="logo">TRAVELR</h1>
        <NavBar planets={planets} moons={moons} bodies={bodies} />
        <Switch>
          <Route exact path='/' render={() => <Home key='home' url={currentUrl} time={arrivalTime} userInfo={userInfo} />} />
          <Route path='/objects/:id' render={({ match }) => {
            const { id } = match.params
            const info = objects.find(obj => obj.object_name.toLowerCase() === id)

            return <Display key={id} info={info} />
          }} />
          <Route path='/moons/:id' render={({ match }) => {
            const { id } = match.params
            const info = objects.find(moon => moon.object_name.toLowerCase() === id)

            return <Display key={id} info={info} />
          }} />
          <Route path="/bodies/:id" render={({ match }) => {
            const { id } = match.params
            const info = objects.find(body => body.object_name.toLowerCase() === id)

            return <Display key={id} info={info} />
          }} />
          <Route component={NotFound} />
        </Switch>
      </div>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  setArrivalTime: (time) => dispatch(setArrivalTime(time)),
  handleObjects: (url) => dispatch(handleObjects(url)),
  handleImages: (url, actionToDispatch) => dispatch(handleImages(url, actionToDispatch))
})

export const mapStateToProps = (state) => ({
  objects: state.objects,
  images: state.images,
  arrivalTime: state.arrivalTime,
  userInfo: state.userInfo,
})

App.propTypes = {
  arrivalTime: PropTypes.number,
  handleObjects: PropTypes.func,
  handleImages: PropTypes.func,
  objects: PropTypes.array,
  setArrivalTime: PropTypes.func,
  userInfo: PropTypes.object,
}

App.defaultProps = {
  arrivalTime: 0,
  handleObjects: [],
  handleImages: [],
  planets: [],
  userInfo: { userDate: "", elapsedDays: 0 },
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(App));