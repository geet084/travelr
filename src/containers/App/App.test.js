import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';
import { shallow, mount } from 'enzyme';
import { App, mapStateToProps, mapDispatchToProps } from './App';
import Display from '../../containers/Display/Display';
import Home from '../../containers/Home/Home';
import { rootReducer } from '../../reducers/index';
import * as actions from '../../actions/';
import moment from 'moment';

describe('App', () => {
  let wrapper;
  let mockProps;

  beforeEach(() => {
    mockProps = {
      images: {
        apod: { media_type: 'image', url: 'http:/someurl' },
        currentImage: {}
      },
      stars: [{}],
      planets: [{}],
      moons: [{}],
      bodies: [{}],
      arrivalTime: { _d: '117' },
      userInfo: { userDate: '10-1-2018', elapsedDays: 0 },
      handleObjects: jest.fn(),
      handleImages: jest.fn(),
      setArrivalTime: jest.fn(),
      setBodies: jest.fn(),
    };

    wrapper = shallow(<App {...mockProps} />);
  });

  describe('initial state', () => {
    mockProps = {
      images: {
        apod: {},
        currentImage: { href: '' }
      },
      stars: [],
      planets: [],
      moons: [],
      bodies: [],
      arrivalTime: {},
      userInfo: { userDate: '', elapsedDays: 0 },
      handleObjects: jest.fn(),
      handleImages: jest.fn(),
      setArrivalTime: jest.fn(),
      setBodies: jest.fn(),
    };

    it('renders without crashing', () => {
      const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk)))
      const div = document.createElement('div');
      ReactDOM.render(
        <Provider store={store}>
          <BrowserRouter>
            <App {...mockProps} />
          </BrowserRouter>
        </Provider>
        , div);
      ReactDOM.unmountComponentAtNode(div);
    });

    it('should match initial snapshot with NO data', () => {
      wrapper = shallow(<App {...mockProps} />, { disableLifecycleMethods: true });

      expect(wrapper).toMatchSnapshot();
    });

    it('should load data and set user arrival time when component mounts', () => {
      wrapper = shallow(<App {...mockProps} />, { disableLifecycleMethods: true });
      wrapper.instance().loadData = jest.fn();
      wrapper.instance().setUserArrivalTime = jest.fn();

      wrapper.instance().componentDidMount();

      expect(wrapper.instance().loadData).toHaveBeenCalledTimes(1);
      expect(wrapper.instance().setUserArrivalTime).toHaveBeenCalledTimes(1);
    });
  });

  describe('loadData', () => {
    let mockNasaUrl;
    let mockServerUrl;

    beforeEach(() => {
      mockNasaUrl = 'http://nasa.org';
      mockServerUrl = 'http://localhost:3000';
      process.env.REACT_APP_NASA_URL = mockNasaUrl;
      process.env.REACT_APP_SERVER_URL = mockServerUrl;
    });

    afterEach(() => {
      delete process.env.REACT_APP_NASA_URL;
      delete process.env.REACT_APP_SERVER_URL;
    });

    it('should call', () => {
      expect(mockProps.handleImages).toHaveBeenCalledTimes(2);
      expect(mockProps.handleObjects).toHaveBeenCalledTimes(1);
    });

    it('should call handleImages with a nasa url and handleApodImage action', () => {
      wrapper.instance().loadData();

      expect(mockProps.handleImages).toHaveBeenCalledWith(mockNasaUrl, actions.handleApodImage);
    });

    it('should call handleObjects with a server url + "/objects"', () => {
      wrapper.instance().loadData();

      expect(mockProps.handleObjects).toHaveBeenCalledWith(mockServerUrl + '/objects');
    });

    it('should call handleImages with a server url + "/images"', () => {
      wrapper.instance().loadData();

      expect(mockProps.handleImages).toHaveBeenCalledWith(mockServerUrl + '/images', actions.handleObjectImages);
    });
  });

  describe('setUserArrivalTime', () => {
    const mockArrivalTime = moment('Mon, 1 Oct 2018 9:00 AM', 'ddd, DD MMM YYYY HH:mm:ss ZZ');
    moment.now = jest.fn().mockReturnValue(mockArrivalTime);

    it('should set the time when the user first arrives on the site', () => {
      expect(mockProps.setArrivalTime).toHaveBeenCalledTimes(1);
    });
  });

  describe('setApodUrl', () => {
    it('should match snapshot with good url', () => {
      wrapper = shallow(<App {...mockProps} />)
      expect(wrapper).toMatchSnapshot();
    });

    it('should match snapshot with bad url', () => {
      let badMockProps = mockProps;
      badMockProps.images.apod.media_type = 'video';

      wrapper = shallow(<App {...badMockProps} />);

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('buildRouteWith', () => {
    it('should build a route for moon objects', () => {
      const mockPath = '/moons/:id';
      const result = wrapper.instance().buildRouteWith(mockPath);

      expect(result.props.path).toEqual(mockPath);
    });
  });

  describe('buildNavBar', () => {
    it('should build a nav bar with just the names of each object', () => {
      mockProps.stars = [{ name: 'Sun', id: 3, size: 'large' }];
      mockProps.planets = [{ name: 'Earth', id: 9 }];
      mockProps.moons = [{ name: 'Luna', id: 1 }, { name: 'Titan', id: 6 }];
      mockProps.bodies = [{ name: 'Asteroid Belt', id: 18 }];
      wrapper = shallow(<App {...mockProps} />);

      const expected = {
        stars: ['Sun'],
        planets: ['Earth'],
        moons: ['Luna', 'Titan'],
        bodies: ['Asteroid Belt']
      };

      const result = wrapper.instance().buildNavBar();

      expect(result.props).toEqual(expected);
    });
  });

  describe('routes', () => {
    const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk)))
    it('should render Home when at the root route', () => {
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter initialEntries={['/']}>
            <App {...mockProps} />
          </MemoryRouter>
        </Provider>
      );

      expect(wrapper.find(Home)).toHaveLength(1);
    });

    it('should render a specific object when on that route', () => {
      mockProps.stars = [{ name: 'Sun' }];
      const wrapper = mount(
        <Provider store={store}>
          <MemoryRouter initialEntries={['/stars/sun']}>
            <App {...mockProps} />
          </MemoryRouter>
        </Provider>
      );

      expect(wrapper.find(Display)).toHaveLength(1);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should mapDispatchToProps for setArrivalTime', () => {
      const mockDispatch = jest.fn();
      const mockTime = 999;
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.setArrivalTime(mockTime);
      expect(mockDispatch).toHaveBeenCalled();
    });

    it('should mapDispatchToProps for handleObjects', () => {
      const mockDispatch = jest.fn();
      const mockUrl = 'mock url';
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.handleObjects(mockUrl);
      expect(mockDispatch).toHaveBeenCalled();
    });

    it('should mapDispatchToProps for handleImages', () => {
      const mockDispatch = jest.fn();
      const mockUrl = 'mock url';
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.handleImages(mockUrl);
      expect(mockDispatch).toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('should mapStateToProps', () => {
      const mockState = {
        images: [{ id: 1 }],
        arrivalTime: 99,
        stars: [{ name: 'sun' }],
        planets: [{ name: 'earth' }, { name: 'mars' }],
        moons: [{ name: 'titan' }, { name: 'moon' }],
        bodies: [{ name: 'asteroid belt' }, { name: 'oort cloud' }],
        userInfo: { arrivalTime: 5, elapsedTime: 88 },
        fakeData1: "should't show up",
        fakeData2: 'stuff'
      };

      const expected = {
        images: [{ id: 1 }],
        arrivalTime: 99,
        stars: [{ name: 'sun' }],
        planets: [{ name: 'earth' }, { name: 'mars' }],
        moons: [{ name: 'titan' }, { name: 'moon' }],
        bodies: [{ name: 'asteroid belt' }, { name: 'oort cloud' }],
        userInfo: { arrivalTime: 5, elapsedTime: 88 },
      };

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });
})
