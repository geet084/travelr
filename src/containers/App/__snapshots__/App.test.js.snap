// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot with bad url 1`] = `
<div
  className="App"
>
  <h1
    className="logo"
  >
    TRAVELR
  </h1>
  <NavBar />
  <Switch>
    <Route
      exact={true}
      path="/"
      render={[Function]}
    />
    <Route
      path="/objects/:id"
      render={[Function]}
    />
    <Route
      component={[Function]}
    />
  </Switch>
</div>
`;

exports[`App should match snapshot with good url 1`] = `
<div
  className="App"
>
  <h1
    className="logo"
  >
    TRAVELR
  </h1>
  <NavBar />
  <Switch>
    <Route
      exact={true}
      path="/"
      render={[Function]}
    />
    <Route
      path="/objects/:id"
      render={[Function]}
    />
    <Route
      component={[Function]}
    />
  </Switch>
</div>
`;
